### PR TITLE
fix race condition causing hang

### LINKED ISSUE #7525: Packer hangs when I run Ansible Provisioner or ansible-playbook via shell provisioner.
##### SUMMARY
Packer hangs when I run Ansible in verbose mode. This happens with ansible provisioner and shell provisioner.

##### ANSIBLE + Packer VERSION

```
==> googlecompute: Provisioning with shell script: /tmp/packer-shell165541510
    googlecompute: ansible 2.7.10
    googlecompute:   config file = None
    googlecompute:   configured module search path = ['/home/packer/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
    googlecompute:   ansible python module location = /usr/local/lib/python3.6/dist-packages/ansible
    googlecompute:   executable location = /usr/local/bin/ansible
    googlecompute:   python version = 3.6.7 (default, Oct 22 2018, 11:32:17) [GCC 8.2.0]

upo@helper-instance:~/ops-infra/packer$ packer --version
1.4.0
```
##### CONFIGURATION

- ansible playbook
```
---
- hosts: localhost
  become: yes
  tasks:

  - name: create Jenkins user
    user: 
      name: jenkins
      state: present
      shell: /bin/bash

  - name: Jenkins | Add Jenkins APT KEY
    apt_key:
      url: https://pkg.jenkins.io/debian/jenkins.io.key 
      state: present

  - name: Jenkins | Add repo to APT sources
    shell: "echo deb https://pkg.jenkins.io/debian-stable binary/ > /etc/apt/sources.list.d/jenkins.list"

  - name: Manual Apt update
    apt: 
      update_cache: yes

  - name: Install dependencies
    package:  
      name:
          - nginx
          - openjdk-8-jdk
          - make
          - git
    
  - name: Install Jenkins
    package:
      name: jenkins
      state: present

```

- packer.json

```
{
    "variables": {
        "zone": "europe-west2-b",
        "project_id": "PID",
        "region": "europe-west2"
      },

    "builders": [
      {
        "type": "googlecompute",
        "region": "{{user `region`}}",
        "zone": "{{user `zone`}}",
        "project_id" : "{{user `project_id`}}",
        "ssh_username": "packer",
        "source_image_family": "ubuntu-1804-lts",
        "preemptible": "true",
        "image_description": "Jenkins with (Java + Docker)",
        "disk_type": "pd-ssd",
        "disk_size": "10",
        "metadata": [
              {

              "block-project-ssh-keys":"false",
              "startup-script-url": "gs://pid-artefacts/init.sh"
              
              }],
        "image_name": "ubuntu1804-jenkins-{{isotime | clean_image_name}}",
	      "image_family": "ubuntu1804-jenkins"

      }
    ],
    

    "provisioners": [
        {
          "type": "file",
          "source": "../ansible/jenkins.yml",
          "destination": "/tmp/jenkins.yml"
        },
        
        {
          "type": "shell",
          "inline": [" ansible --version && cd /tmp && ansible-playbook jenkins.yml -vvv"],
          "pause_before": "140s"
        }
      ]
  } 
  
```

##### OS / ENVIRONMENT
Google Compute Engine + Ubuntu 18.04 LTS

##### STEPS TO REPRODUCE
Build a packer image.

##### EXPECTED RESULTS

Here is a log without the verbose flag.

```
googlecompute output will be in this color.

==> googlecompute: Checking image does not exist...
==> googlecompute: Creating temporary SSH key for instance...
==> googlecompute: Using image: ubuntu-1804-bionic-v20190404
==> googlecompute: Creating instance...
    googlecompute: Loading zone: europe-west2-b
    googlecompute: Loading machine type: n1-standard-1
    googlecompute: Requesting instance creation...
    googlecompute: Waiting for creation operation to complete...
    googlecompute: Instance has been created!
==> googlecompute: Waiting for the instance to become running...
    googlecompute: IP: 35.230.134.102
==> googlecompute: Using ssh communicator to connect: 35.230.134.102
==> googlecompute: Waiting for SSH to become available...
==> googlecompute: Connected to SSH!
==> googlecompute: Uploading ../ansible/jenkins.yml => /tmp/jenkins.yml
jenkins.yml 741 B / 741 B [=============================================================================================================================================================================================================] 100.00% 0s
==> googlecompute: Pausing 2m20s before the next provisioner...
==> googlecompute: Provisioning with shell script: /tmp/packer-shell487567343
    googlecompute: ansible 2.7.10
    googlecompute:   config file = None
    googlecompute:   configured module search path = ['/home/packer/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
    googlecompute:   ansible python module location = /usr/local/lib/python3.6/dist-packages/ansible
    googlecompute:   executable location = /usr/local/bin/ansible
    googlecompute:   python version = 3.6.7 (default, Oct 22 2018, 11:32:17) [GCC 8.2.0]
==> googlecompute:  [WARNING]: Unable to parse /etc/ansible/hosts as an inventory source
==> googlecompute:  [WARNING]: No inventory was parsed, only implicit localhost is available
==> googlecompute:  [WARNING]: provided hosts list is empty, only localhost is available. Note
==> googlecompute: that the implicit localhost does not match 'all'
    googlecompute: PLAY [localhost] ***************************************************************
    googlecompute: TASK [Gathering Facts] *********************************************************
    googlecompute: ok: [localhost]
    googlecompute: TASK [create Jenkins user] *****************************************************
    googlecompute: changed: [localhost]
    googlecompute: TASK [Jenkins | Add Jenkins APT KEY] *******************************************
    googlecompute: changed: [localhost]
    googlecompute: TASK [Jenkins | Add repo to APT sources] ***************************************
    googlecompute: changed: [localhost]
    googlecompute: TASK [Manual Apt update] *******************************************************
    googlecompute: changed: [localhost]
    googlecompute: TASK [Install dependencies] ****************************************************
    googlecompute: changed: [localhost]
    googlecompute: TASK [Install Jenkins] *********************************************************
    googlecompute: changed: [localhost]
    googlecompute: PLAY RECAP *********************************************************************
    googlecompute: localhost                  : ok=7    changed=6    unreachable=0    failed=0
==> googlecompute: Deleting instance...
    googlecompute: Instance has been deleted!
==> googlecompute: Creating image...
==> googlecompute: Deleting disk...
    googlecompute: Disk has been deleted!
Build 'googlecompute' finished.

==> Builds finished. The artifacts of successful builds are:
--> googlecompute: A disk image was created: ubuntu1804-jenkins-2019-04-18t09-19-32z

```

I filed an issue with ansible but it works fine in with a startup script as I demonstrated in the case .
https://github.com/ansible/ansible/issues/55510

### LINKED ISSUE #7529: Packer 1.4.0 hangs while installing ansible-2.7.10 into a python virtualenv using pip
**SUMMARY:**
Packer 1.4.0 hangs while installing ansible-2.7.10 into a python virtualenv using pip, the Packer EBS builder, and shell provisioner. The same template and builder works on Packer 1.3.5. This is on Windows 10 64-bit. 

**SAMPLE TEMPLATE/LOGS:**
https://gist.github.com/AndrewCi/3d768238412885a8e48cf4f4a58e56ea

### PR DESCRIPTION
**Pull request by SwampDragons**
Something about the iochan.LineReader was breaking the way we read whether a channel is closed. I reverted to our old way of piping and scrubbing stderr and stdout to end this race condition.

Closes #7525
Closes #7529

### CODE REVIEW

**Comment by SwampDragons** [comment]
Windows build: [packer.zip](https://github.com/hashicorp/packer/files/3123244/packer.zip)

linux build: 
[packer.zip](https://github.com/hashicorp/packer/files/3123245/packer.zip)

osx build:
[packer.zip](https://github.com/hashicorp/packer/files/3123252/packer.zip)

**Comment by AndrewCi** [comment]
Confirming this fixes my issue #7529 on Windows 10.

**Comment by azr** [comment]
Oh, very good catch, I think the fix might come from re-adding that part :

```
	// Start the goroutine to watch for the exit
	go func() {
		defer stdout_w.Close()
		defer stderr_w.Close()
		r.Wait()
	}()
```

I really believed it would not be necessary since the command would close this upon termination; and my test wasn't thorough enough.
🤦‍♂ my bad !

File: packer/rpc/communicator.go [hunk 1]
```diff
@@ -101,10 +101,13 @@ func (c *communicator) Start(ctx context.Context, cmd *packer.RemoteCmd) (err er
 
 	go func() {
 		conn, err := c.mux.Accept(responseStreamId)
+		log.Printf("Megan waiting...")
```
Comment (azr) [inline packer/rpc/communicator.go]: woops ! 🙂

**Comment by SwampDragons** [comment]
Yeah, that goroutine you call about above was necessary but not sufficient; even re-adding it, using the LineReader with the scanner was causing trouble. There were two entangled races, which made it extra fun :)

**Pull request merged**